File: Java/Queue/src/Printer.java
import org.jetbrains.annotations.NotNull;

import java.util.PriorityQueue;

public class Printer {

    private final PriorityQueue<Document> docs = new PriorityQueue<Document>();

    public Printer(){
    }

    public void addToQueue(Document... args){
        for(int i = 0; i <= args.length-1; i++){
            System.out.println("Adding " + args[i].getName() + " to the printer queue...");
            docs.add(args[i]);
        }
        System.out.println("All files added to the queue");
    }

    public boolean printDocument(){
        while(!docs.isEmpty()){
            System.out.println("Printing " + docs.remove().getName() + "..." );
            if(docs.isEmpty()){
                System.out.println("There is no documents anymore to be printed");
                return true;
            }
        }
        System.out.println("There is no documents to be printed");
        return false;
    }
}

File: Kotlin/Queue/src/exercise1/Printer.kt
package exercise1

import java.util.PriorityQueue
import kotlinx.coroutines.*
import kotlinx.coroutines.Dispatchers
import kotlin.system.measureTimeMillis

class Printer {

    private val pQueue = PriorityQueue<Document>()

    suspend fun addToQueue(vararg documents: Document) {
        runBlocking(Dispatchers.IO) {
            val queuePreparation: Deferred<Boolean> =
                async {
                    documents.forEach {
                        println("Adding ${it.name} to print queue...")
                        pQueue.add(it)
                        delay(2000)
                    }
                    true
                }

            if (queuePreparation.await()) {
                documents.forEach {
                    println("exercise1.Document: ${it.name} is ready to be printed")
                }
            }
        }
    }

    suspend fun printDocument() : Boolean {
        var status: Boolean = false
        runBlocking(Dispatchers.IO) {
            val timer = measureTimeMillis {
                val printProcess: Deferred<Boolean> = async<Boolean>(context = Dispatchers.IO) {
                    repeat(pQueue.size, {
                        try {
                            val document = pQueue.remove()
                            println("Printing ${document.name}")
                            println("Total pages: ${document.pages}")
                            var page = document.pages
                            repeat(document.pages){
                                println("${document.name} | page $page")
                                delay(5000)
                                page--
                            }

                        }catch (e: NoSuchElementException){
                            println("exercise1.Printer queue is empty")
                        }
                    })
                    true
                }
               status = printProcess.await()
            }
            println("Process time: $timer")
        }
        return status
    }
}